File: backend/app/models/contract.py
"""
Dasfax — Tier 3 shared contract.

THIS FILE IS THE CONTRACT for the two halves of Tier 3:

    WS5 (claim extraction + evidence retrieval)  --produces-->  ClaimExtractionResult
    WS6 (assessment)                             --consumes-->  ClaimExtractionResult
                                                 --produces-->  Assessment (one per Claim)

It is owned by WS5 by default (recon found no types, no backend, no mock anywhere in
the repo). WS6 and WS3 should import these models rather than redefining shapes, so the
seam stays frozen. If a field needs to change, change it HERE and tell WS6/WS3 — do not
fork a parallel definition.

Field ownership is annotated inline:
    [WS5] populated by WS5.        [WS6] populated by WS6.        [WS3] envelope/routing.

Pydantic v2.
"""
from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, Field, field_validator


# --------------------------------------------------------------------------- #
# Enums
# --------------------------------------------------------------------------- #
class ClaimType(StrEnum):
    """What kind of statement a sentence is. WS5 only checks FACTUAL claims;
    OPINION / PREDICTION are extracted so WS6 can surface an 'Opinion' treatment
    without WS5 wasting retrieval budget on them."""

    FACTUAL = "factual"          # a verifiable, checkable assertion about the world
    OPINION = "opinion"          # value judgement / subjective statement
    PREDICTION = "prediction"    # claim about the future, not yet checkable


class AssessmentStatus(StrEnum):
    """WS6 verdict vocabulary (proposal §2.3). Defined here so the whole Tier 3
    contract lives in one file; WS5 never sets these."""

    SUPPORTED = "supported"
    PARTIALLY_SUPPORTED = "partially_supported"
    CONTRADICTED = "contradicted"
    NEEDS_REVIEW = "needs_review"      # insufficient evidence -> honest fallback
    OPINION = "opinion"               # not a factual claim; nothing to verify


# --------------------------------------------------------------------------- #
# Inputs (what WS3/WS1 hand to WS5)
# --------------------------------------------------------------------------- #
class ArticleInput(BaseModel):
    """Cleaned article text handed to WS5. Text is expected to be already
    extracted/cleaned upstream (WS1 article extraction). WS5 does not fetch pages."""

    url: str = Field(..., description="Canonical URL of the article (for logging/citation dedup).")
    title: str | None = Field(None, description="Article headline, if available.")
    text: str = Field(..., min_length=1, description="Cleaned article body text.")
    lang: str | None = Field(None, description="BCP-47 language tag, e.g. 'en'. Advisory only.")
    source_domain: str | None = Field(None, description="Hostname, e.g. 'straitstimes.com'.")
    published_at: datetime | None = Field(None, description="Publish timestamp if known.")

    @field_validator("text")
    @classmethod
    def _text_not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("text must not be blank")
        return v


# --------------------------------------------------------------------------- #
# Evidence + Claim (WS5 output)
# --------------------------------------------------------------------------- #
class Evidence(BaseModel):
    """One retrieved snippet that bears on a claim. [WS5]"""

    snippet: str = Field(..., description="Short retrieved passage relevant to the claim.")
    source_url: str = Field(..., description="Resolvable URL the snippet came from.")
    source_title: str | None = Field(None, description="Title of the source page.")
    source_domain: str | None = Field(None, description="Hostname of the source.")
    published_at: datetime | None = Field(None, description="Source publish date if known.")
    relevance_score: float | None = Field(
        None, ge=0.0, le=1.0,
        description="Retrieval relevance in [0,1]. Advisory; WS6 may re-rank.",
    )


class Claim(BaseModel):
    """A single extracted claim with its retrieved evidence.

    WS5 fills everything here. WS6 reads `text` + `evidence` and writes a separate
    Assessment keyed by `id` — WS6 does NOT mutate the Claim.
    """

    id: str = Field(..., description="Stable id within one article (e.g. 'c1'). Assessment.claim_id references this.")
    text: str = Field(..., description="The claim as a self-contained sentence. WS2 fuzzy-matches this back onto the DOM.")
    claim_type: ClaimType = Field(..., description="factual / opinion / prediction.")
    checkworthiness: float = Field(
        ..., ge=0.0, le=1.0,
        description="How load-bearing/verifiable this claim is. Used for ranking; higher = check first.",
    )
    rank: int = Field(..., ge=1, description="1 = most important claim to check. Only ranked FACTUAL claims carry evidence.")
    search_query: str | None = Field(None, description="Query WS5 issued to retrieve evidence (for transparency/debug).")
    evidence: list[Evidence] = Field(default_factory=list, description="Retrieved snippets. Empty for non-factual or unretrieved claims.")

    # --- Anchoring hints (WS5 -> WS2) -------------------------------------- #
    # Offsets index ArticleInput.text — the same cleaned body text WS1 produces as
    # `bodyText` — so WS2 can map the span onto the live DOM:
    # `ArticleInput.text[char_start:char_end]` is the source span this claim came from.
    # All four are None when WS5 could not place the claim confidently; WS2 then falls
    # back to fuzzy matching on `text`.
    char_start: int | None = Field(None, ge=0, description="Start offset of the claim's source span within ArticleInput.text.")
    char_end: int | None = Field(None, ge=0, description="End offset (exclusive) of the source span within ArticleInput.text.")
    prefix: str | None = Field(None, description="Up to 32 chars of ArticleInput.text immediately BEFORE the span (disambiguates repeats).")
    suffix: str | None = Field(None, description="Up to 32 chars of ArticleInput.text immediately AFTER the span.")


# --------------------------------------------------------------------------- #
# WS5 result envelope (WS5 -> WS6)
# --------------------------------------------------------------------------- #
class ExtractionStats(BaseModel):
    """Observability for WS5. Lets the pitch quote real numbers and lets WS3 log tiers."""

    total_claims_extracted: int = 0
    factual_claims: int = 0
    dropped_opinion_or_prediction: int = 0
    dropped_duplicate: int = 0
    kept_after_ranking: int = 0
    claims_with_evidence: int = 0
    claims_anchored: int = 0


class ClaimExtractionResult(BaseModel):
    """WS5's deliverable. This is exactly what WS6 receives."""

    url: str
    title: str | None = None
    claims: list[Claim] = Field(default_factory=list, description="Ranked, deduped. At most `max_claims` factual claims carry evidence.")
    stats: ExtractionStats = Field(default_factory=ExtractionStats)
    model_meta: dict = Field(
        default_factory=dict,
        description="Freeform: which LLM/search backend + version produced this (mock vs real).",
    )


# --------------------------------------------------------------------------- #
# Assessment (WS6 output) — defined here so the seam is visible; WS5 never writes it.
# --------------------------------------------------------------------------- #
class Citation(BaseModel):
    """A specific piece of evidence WS6 leaned on for its verdict. [WS6]"""

    snippet: str
    source_url: str
    source_title: str | None = None


class Assessment(BaseModel):
    """WS6's verdict for one claim. [WS6] — included in the contract for completeness.

    Rule WS6 must honour (proposal §2.5): every non-OPINION assessment carries at least
    one Citation, or its status is NEEDS_REVIEW. Malformed output must never reach the client.
    """

    claim_id: str = Field(..., description="References Claim.id.")
    status: AssessmentStatus
    explanation: str = Field(..., description="Plain-language reason, grounded in the citations.")
    confidence: float | None = Field(None, ge=0.0, le=1.0)
    citations: list[Citation] = Field(default_factory=list)


# Convenience: the shape WS3 would return to the extension (WS2) once WS6 has run.
class VerifiedClaim(BaseModel):
    """Claim + its Assessment, joined by WS3 for the client. Neither WS5 nor WS6 build this alone."""

    claim: Claim
    assessment: Assessment | None = None


# --------------------------------------------------------------------------- #
# Client envelope (backend -> WS2)
#
# MIRRORS src/shared/contract.ts, where WS2 authored these shapes first as its
# own "client envelope". Now that the backend actually produces them, THIS file
# is the source of truth for the envelope too — Darren / Terry, sync
# src/shared/contract.ts from here and drop its "proposed to WS3" caveat.
#
# Field names are camelCase ON PURPOSE: they are the literal wire format WS2's
# isAnalysisResponse() guard checks. Renaming them to snake_case breaks the
# renderer silently (the guard returns false and the response is dropped).
# --------------------------------------------------------------------------- #
class AnalysisStatus(StrEnum):
    """WS2's `AnalysisStatus` union, verbatim."""

    COMPLETE = "complete"
    PROCESSING = "processing"
    FAILED = "failed"
    SKIPPED = "skipped"      # Tier 0 trusted source -> WS2 renders the "trusted" pill


class ArticleVerdictLevel(StrEnum):
    """WS2's `ArticleVerdictLevel` union, verbatim.

    OK/CAUTION/HIGH_RISK are Tier 3 rollups: `article_verdict_for()` derives them from
    the per-claim verdicts and nothing else may produce OK — it means "claims were
    checked and at least one holds up". UNRATED is the neutral state for every path that
    ends before a Tier 3 rollup: no article text, zero checkable claims, or a Tier 2
    screen that cleared the page without escalating. It is not a clean bill of health and
    not a warning; the reason for it rides in `ArticleVerdict.summary`. TRUSTED is Tier 0
    (whitelisted domain). See app/services/envelope.py.
    """

    TRUSTED = "trusted"
    OK = "ok"
    CAUTION = "caution"
    HIGH_RISK = "high_risk"
    UNRATED = "unrated"


class ArticleVerdict(BaseModel):
    """Article-level rollup shown on WS2's summary pill and at the top of the panel."""

    level: ArticleVerdictLevel
    summary: str
    confidence: float | None = Field(None, ge=0.0, le=1.0)


class AnalysisError(BaseModel):
    """Non-fatal problem to surface alongside a still-valid envelope."""

    code: str
    message: str


class AnalysisRequest(BaseModel):
    """What the extension POSTs to /analyze.

    `text` is OPTIONAL so WS3 can wire and test the fetch before WS1 forwards
    extracted body text; the endpoint answers with a valid empty envelope until it
    arrives, and lights up on its own once it does.
    """

    url: str = Field(..., description="Canonical URL of the page being analyzed.")
    title: str | None = Field(None, description="Page/article title, if the content script has one.")
    text: str | None = Field(None, description="Cleaned article body text (WS1 `bodyText`). Absent until WS1 forwards it.")
    images: list[str] | None = Field(None, description="Article image URLs. Unused by WS5 today; carried for WS6.")


class AnalysisResponse(BaseModel):
    """The client-facing envelope WS2 renders. Must satisfy isAnalysisResponse()."""

    schemaVersion: Literal["1.0"] = "1.0"
    url: str
    status: AnalysisStatus
    articleVerdict: ArticleVerdict
    verifiedClaims: list[VerifiedClaim] = Field(default_factory=list)
    # WS2 types this `errors?: AnalysisError[]`; `null` is not a member of that type,
    # so default to an empty list. Do not switch this model to exclude_none — that
    # would also drop `assessment: null`, which WS2's isVerifiedClaim() requires.
    errors: list[AnalysisError] = Field(default_factory=list)

File: src/shared/contract.ts
/**
 * TS mirror of the frozen Tier 3 contract.
 *
 * SOURCE OF TRUTH: backend/app/models/contract.py (Pydantic v2, owned by WS5).
 * Keep these shapes in sync by hand — there is no codegen. Only the fields WS2
 * actually renders are mirrored here; the backend model has more (stats,
 * model_meta, search_query provenance, …) that the in-page UI does not need.
 *
 * `AnalysisResponse` at the bottom is WS2-owned: the thin client envelope that
 * wraps `VerifiedClaim[]` for the summary pill. WS3 will either ratify it or
 * hand us theirs; it must wrap `VerifiedClaim` unchanged.
 */

// --- mirror of contract.py -------------------------------------------------- //

export type ClaimType = "factual" | "opinion" | "prediction";

export type AssessmentStatus =
  | "supported"
  | "partially_supported"
  | "contradicted"
  | "needs_review"
  | "opinion";

export interface Evidence {
  snippet: string;
  source_url: string;
  source_title?: string | null;
  source_domain?: string | null;
  relevance_score?: number | null;
}

export interface Claim {
  id: string;
  /** A self-contained sentence. This is the ONLY handle WS2 has for re-anchoring
   * the claim onto the live DOM — see src/content/anchor/. */
  text: string;
  claim_type: ClaimType;
  checkworthiness: number;
  rank: number;
  search_query?: string | null;
  evidence: Evidence[];
  /** OPTIONAL anchoring hints. Not in the contract yet — requested from WS5.
   * `char_start`/`char_end` are offsets of `text` within the cleaned article
   * body; `prefix`/`suffix` are short context windows. The matcher uses them
   * when present and works without them. */
  char_start?: number | null;
  char_end?: number | null;
  prefix?: string | null;
  suffix?: string | null;
}

export interface Citation {
  snippet: string;
  source_url: string;
  source_title?: string | null;
}

export interface Assessment {
  claim_id: string;
  status: AssessmentStatus;
  explanation: string;
  confidence?: number | null;
  citations: Citation[];
}

export interface VerifiedClaim {
  claim: Claim;
  assessment: Assessment | null;
}

// --- mirror of screening.py (Tier 2, owned by WS4) ------------------------- //

/**
 * What WS3's background worker sends to `POST /tier2/screen`.
 *
 * `images` is the wire name the content script already uses; the backend accepts
 * both `images` and its own `image_urls` spelling. Mirrored here so the two names
 * cannot silently diverge again — an unaliased mismatch is invisible, because
 * Pydantic drops unknown keys and every page would then screen image-clean.
 *
 * `text` is optional on purpose: Readability fails on some pages, and Tier 2 can
 * still screen the images when it does.
 */
export interface ScreeningRequest {
  url: string;
  title?: string;
  text?: string;
  images?: string[];
  published_at?: string;
  source_domain?: string;
}

export interface ImageScreeningResult {
  image_url: string;
  is_synthetic_score: number;
  flagged: boolean;
  reason?: string | null;
}

/**
 * Tier 2 is a ROUTER, not a verdict. Nothing here is rendered to the reader —
 * WS3 routes on `escalate_to_tier3` and the user only ever sees Tier 3 output.
 */
export interface ScreeningResult {
  url: string;
  text_score: number;
  text_flagged: boolean;
  image_results: ImageScreeningResult[];
  max_image_score: number;
  image_flagged: boolean;
  escalate_to_tier3: boolean;
  reasons: string[];
  /** A component fell back or timed out; treat the decision as lower-confidence. */
  degraded: boolean;
  /** False when there was no usable body text and the text model never ran. */
  text_scored: boolean;
  images_timed_out: boolean;
  latency_ms: number;
  within_latency_budget: boolean;
  model_meta: Record<string, unknown>;
}

// --- WS2-owned client envelope (proposed to WS3) --------------------------- //

export type AnalysisStatus = "complete" | "processing" | "failed" | "skipped";
export type ArticleVerdictLevel = "trusted" | "ok" | "caution" | "high_risk";

export interface ArticleVerdict {
  level: ArticleVerdictLevel;
  summary: string;
  /** `null` when unset — the backend serialises nulls rather than omitting keys,
   * because `VerifiedClaim.assessment` must stay an explicit `null` for the guard. */
  confidence?: number | null;
}

export interface AnalysisError {
  code: string;
  message: string;
}

/** Why the tier cascade did what it did. Optional and advisory — useful for
 * demoing that most pages stop at Tier 2, which is the whole point of the
 * cascade and is otherwise invisible. Never required by `isAnalysisResponse`. */
export interface Tier2Summary {
  escalated: boolean;
  textScore: number;
  maxImageScore: number;
  reasons: string[];
  degraded: boolean;
  latencyMs: number;
}

export interface AnalysisResponse {
  schemaVersion: "1.0";
  url: string;
  /** `skipped` == Tier 0 trusted source; render the "trusted" pill, no claims.
   * NOTE: there is deliberately no status meaning "Tier 2 screened this and
   * stopped". A non-escalated page returns `complete` with
   * `articleVerdict.level: "ok"` and no claims — see backend/app/pipeline/analyze.py.
   * WS2/WS3 should ratify that or add a status. */
  status: AnalysisStatus;
  articleVerdict: ArticleVerdict;
  verifiedClaims: VerifiedClaim[];
  errors?: AnalysisError[] | null;
  tier2?: Tier2Summary | null;
}

// --- runtime guards -------------------------------------------------------- //

const CLAIM_TYPES = new Set<ClaimType>(["factual", "opinion", "prediction"]);
const ASSESSMENT_STATUSES = new Set<AssessmentStatus>([
  "supported",
  "partially_supported",
  "contradicted",
  "needs_review",
  "opinion",
]);
const ANALYSIS_STATUSES = new Set<AnalysisStatus>([
  "complete",
  "processing",
  "failed",
  "skipped",
]);
const VERDICT_LEVELS = new Set<ArticleVerdictLevel>([
  "trusted",
  "ok",
  "caution",
  "high_risk",
]);

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null;
}

function isEvidence(v: unknown): v is Evidence {
  return isRecord(v) && typeof v.snippet === "string" && typeof v.source_url === "string";
}

function isCitation(v: unknown): v is Citation {
  return isRecord(v) && typeof v.snippet === "string" && typeof v.source_url === "string";
}

function isClaim(v: unknown): v is Claim {
  return (
    isRecord(v) &&
    typeof v.id === "string" &&
    typeof v.text === "string" &&
    typeof v.claim_type === "string" &&
    CLAIM_TYPES.has(v.claim_type as ClaimType) &&
    typeof v.rank === "number" &&
    Array.isArray(v.evidence) &&
    v.evidence.every(isEvidence)
  );
}

function isAssessment(v: unknown): v is Assessment {
  return (
    isRecord(v) &&
    typeof v.claim_id === "string" &&
    typeof v.status === "string" &&
    ASSESSMENT_STATUSES.has(v.status as AssessmentStatus) &&
    typeof v.explanation === "string" &&
    Array.isArray(v.citations) &&
    v.citations.every(isCitation)
  );
}

function isVerifiedClaim(v: unknown): v is VerifiedClaim {
  return (
    isRecord(v) &&
    isClaim(v.claim) &&
    (v.assessment === null || isAssessment(v.assessment))
  );
}

function isImageScreeningResult(v: unknown): v is ImageScreeningResult {
  return (
    isRecord(v) &&
    typeof v.image_url === "string" &&
    typeof v.is_synthetic_score === "number" &&
    typeof v.flagged === "boolean"
  );
}

/**
 * Structural check for a Tier 2 response. WS3 routes on `escalate_to_tier3`, so a
 * malformed screen must fail loudly rather than be read as "nothing to see here" —
 * a missing boolean coerces to false, which would silently drop the article.
 */
export function isScreeningResult(v: unknown): v is ScreeningResult {
  return (
    isRecord(v) &&
    typeof v.url === "string" &&
    typeof v.text_score === "number" &&
    typeof v.text_flagged === "boolean" &&
    typeof v.image_flagged === "boolean" &&
    typeof v.escalate_to_tier3 === "boolean" &&
    typeof v.text_scored === "boolean" &&
    typeof v.degraded === "boolean" &&
    Array.isArray(v.reasons) &&
    v.reasons.every((r) => typeof r === "string") &&
    Array.isArray(v.image_results) &&
    v.image_results.every(isImageScreeningResult)
  );
}

/**
 * Structural check that untrusted backend output matches `AnalysisResponse`.
 * Malformed responses must never reach the renderer (proposal §2.5).
 */
export function isAnalysisResponse(v: unknown): v is AnalysisResponse {
  if (!isRecord(v)) return false;
  if (v.schemaVersion !== "1.0") return false;
  if (typeof v.url !== "string") return false;
  if (typeof v.status !== "string" || !ANALYSIS_STATUSES.has(v.status as AnalysisStatus)) {
    return false;
  }
  if (!isRecord(v.articleVerdict)) return false;
  if (
    typeof v.articleVerdict.level !== "string" ||
    !VERDICT_LEVELS.has(v.articleVerdict.level as ArticleVerdictLevel)
  ) {
    return false;
  }
  if (typeof v.articleVerdict.summary !== "string") return false;
  if (!Array.isArray(v.verifiedClaims) || !v.verifiedClaims.every(isVerifiedClaim)) {
    return false;
  }
  return true;
}
